### json 불러와서 저장하기

In [1]:
import numpy as np
import pandas as pd
import json
import jpype
from bs4 import BeautifulSoup as bs

data_path = "data"

json_train=[]
for i in range(0,18):
    json_file = json.loads(open("%s/%d.json" % (data_path, i)).read())
    json_train += json_file["articles"]

### json에서 text 형태소 추출

In [4]:
from konlpy.tag import Mecab

mecab = Mecab()

labeled_train = []

for cnt, article in enumerate(json_train):
    text = bs(article["text"],"lxml").text
    
    if cnt % 10000 == 0:
        print cnt

    labeled_train.append({
            "istroll": article["is_troll"],
            "text_morphs": mecab.morphs(text)
        })

labeled_train = pd.DataFrame.from_dict(labeled_train)
labeled_train

0
10000
20000
30000
40000
50000
60000
70000
80000
90000
100000
110000
120000
130000
140000
150000
160000
170000


istroll                                        text_morphs
0        False  [일간, 워스트, 마을, 회관, 입니다, ., 공지, 글, 은, 이곳, 에, 올라옵...
1        False  [<, 일간, 워스트, >, 입니다, 추천, 은, ', 일, 워, 로, ', ,, ...
2        False                              [벌써, 열린, 거, 에요, !, ?]
3        False                          [아햏햏, 고구마, 장사, 가, 힘들, 어요]
4        False                                   [ㅎ, ㄷ, ㄷ, 사스, 가]
5        False       [민영화, 는, 가, 라, !, 우리, 는, 안녕, 하, 지, 못하, 다, !]
6        False  [제, 가, 친, 드립, 이, 세, 시간, 뒤, 현상계, 에, 구현, 되, 어, 있...
7        False                                     [신장, 팔, 아, 간다]
8        False  [페이스북, 디자인, 카피, 하, 고, 좋, 아요, /, 민영화, 버튼, 달린, 사...
9        False  [일, 웤ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ, 손, 꾸, 락, 으로, 동그...
10       False         [근데, 민영화, 말, 고, 사, 영화, 라고, 부르, 면, 안, 됨, ?]
11       False  [민영화, 라고, 쓰, 면, 민영환, 이, 떠올라서, 골룸, 하, 니, 사, 영화,...
12       False                                      [어떻게, 해야죠, ?]
13       False  [1, ., 일간, 워스트, 개설, 2, ., 일간, 워스트, 개설, 첫날, 트래픽...
14       False                                          [ㅈ, ㄱ, ㄴ]
15       False  [민영화, 했, 습니다, 로, 고쳐, 주, 세여, 뿌, 잉, 뿌, 잉, 는, 공지,...
16       False  [여러분, 의, 치킨, 은, 양념, 하, 십니까, 아침, 부터, 치킨, 이, 땡겨,...
17       False  [아이디, 뭘로, 가입, 했, 는지, 기억, 이, 나, 질, 않, 아서, 고민, 입...
18       False                                               [ㅠㅠ]
19       False  [아이디, 는, 받, 아, 놓, 고, 로그인, 은, 이메일, 로, 함, ., ., ...
20       False  [애국, 진보, 인가, ㅂ, ㅅ, 아치, 까, 는, 대신, 에, 눈, 새, 상마, ...
21       False                                [쏘, 워, 츙워츙워츙, 원, ㅌ]
22       False  [', 기행, 천지, ', 일, 베, …, 편의점, 호빵, ,, 담배, 와, 같이,...
23       False  [약속, 나가, 려다가, 번뜩, 떠올라서, 후딱, 베, 낌, 맘, 에, 드, 시, ...
24       False                                     [외쳐, !, EE, !]
25       False                                                 []
26       False                                [못, 하, 면, ㅁ, ㅇㅎ, .]
27       False  [http, :, //, bbs, 1, ., agora, ., media, ., d...
28       False                                [일, 베, 충, 박멸, !, !]
29       False                                [하, ↘, 지마, 루, 요, ↗]
...        ...                                                ...
172014   False  [얼마, 전, 우리, 빌라, 의, 귀염둥이, 보더, 콜리, 장군, 이, 가, 없, ...
172015    True  [머리, 가, 다, 빠져, 버렸, 는데, 머리, 다시, 자라, 게, 하, 는, 방법...
172016   False                                                 []
172017   False  [전, 아이돌, 성장, 물, 을, 안, 좋아하, 는, 편, 입니다, ., 뻔하, 잖...
172018   False  [저, 사진, 넣, 으려니, 지원, 하, 지, 않, 는, 파일, 이, 라는데, .,...
172019   False  [아까, 부터, 제, 가, 쓴, 글, 이, 밭, 에서, 보이, 지, 도, 않, 고,...
172020   False     [www, ., ilwar, ., com, /, fashion, /, 189620]
172021   False  [트롤, 밭, 게시물, 이, 보이, 기, 시작, 합니다, ., .., http, :...
172022   False  [http, :, //, www, ., segye, ., com, /, conten...
172023   False  [친구, 가, 다른, 친구, 하, 고, 같이, 있, 으면, ,, 바로, 카톡, 으로...
172024   False  [그, 중, 우리, 나라, 에서, 독재, 한, 분, 들, 또는, 논란거리, 를, 마...
172025   False  [무서웠, 습니다, ., ., 마치, 마약, 금단, 증세, +, 외로움, +, 소통...
172026   False                                            [비, 연시]
172027   False  [잘, 쓰, 던, 아이패드, 를, 팔, 고, (, 두, 달, 전, 중고, 로, 샀,...
172028   False  [어떻게, 알, 고, 친, 추, 를, 하, 신지, 모를, 한, 농부, 님, 과, 서...
172029   False  [이르, 지만, 작별, 하, 고, 싶, 네요, ., 2014, 년, 얼마, 남, 지...
172030   False  [올해, 도, 이제, ., .., 26, 시간, 여, 남, 았, 구, 농, ., ....
172031   False  [겜, 덕인, 나농, 이, 마우스, 를, 바꾸, 려고, 하, 는데, 게이밍, 마우스...
172032   False  [오늘, ', 이강윤, 의, 오늘, ', 에서, 우원식, 의원, 이, 인터뷰, 도,...
172033   False                    [카피, 문구, 어디, 서, 본, 듯, 한데, ., .]
172034   False  [때문, 에, 땅콩, 부사장, 님, 은, 곧, 수감, 될, 예정, 입니다, 이상, ...
172035   False  [...., 라고, 쓰, 고, 농약, 짤, 줍, 용, 글, 이, 라고, 읽, 는다,...
172036   False  [국내, 최고, 의, 팟캐스트, ", 그것, 은, 알, 기, 싫, 다, ", 의, ...
172037   False  [허접, 한, 실력, 이, 지만, 개발자, 가, 직업, 인지라, 보, 고, 있, 는...
172038   False  [무한도전, 의, 기획력, 에, 감탄, 하, 는, 것, 이, 한, 두, 번, 이, ...
172039   False  [분명, 어데, 부딪치, 지, 도, 않, 았, 는데, 멍, 이, 들었, 농, ., ...
172040   False  [나농, 이, 어릴, 적, 은, 패스, 하, 고, 나농, 이, 좀, 머리, 가, 자...
172041   False  [외국인, 길드, 들, 었, 다가, 말, 이, 안, 통해서, 나왔, 는데, 외롭, ...
172042   False  [한, 것, 도, 없, 는데, 벌써, 시간, 이, 이렇게

###bag of words

In [5]:
from sklearn.feature_extraction.text import CountVectorizer

max_features = 1000

vectorizer = CountVectorizer(analyzer = "word", tokenizer = None, preprocessor = None,
                            stop_words = None, max_features=max_features)

train_data_features = vectorizer.fit_transform((labeled_train["text_morphs"].apply(lambda x:" ".join(x)))).toarray()

col = ["bow_text_%s" % data for data in vectorizer.get_feature_names()]
labeled_train = pd.concat([labeled_train, pd.DataFrame(train_data_features, columns = col, index=labeled_train.index)],axis=1)

labeled_train

istroll                                        text_morphs  \
0        False  [일간, 워스트, 마을, 회관, 입니다, ., 공지, 글, 은, 이곳, 에, 올라옵...   
1        False  [<, 일간, 워스트, >, 입니다, 추천, 은, ', 일, 워, 로, ', ,, ...   
2        False                              [벌써, 열린, 거, 에요, !, ?]   
3        False                          [아햏햏, 고구마, 장사, 가, 힘들, 어요]   
4        False                                   [ㅎ, ㄷ, ㄷ, 사스, 가]   
5        False       [민영화, 는, 가, 라, !, 우리, 는, 안녕, 하, 지, 못하, 다, !]   
6        False  [제, 가, 친, 드립, 이, 세, 시간, 뒤, 현상계, 에, 구현, 되, 어, 있...   
7        False                                     [신장, 팔, 아, 간다]   
8        False  [페이스북, 디자인, 카피, 하, 고, 좋, 아요, /, 민영화, 버튼, 달린, 사...   
9        False  [일, 웤ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ, 손, 꾸, 락, 으로, 동그...   
10       False         [근데, 민영화, 말, 고, 사, 영화, 라고, 부르, 면, 안, 됨, ?]   
11       False  [민영화, 라고, 쓰, 면, 민영환, 이, 떠올라서, 골룸, 하, 니, 사, 영화,...   
12       False                                      [어떻게, 해야죠, ?]   
13       False  [1, ., 일간, 워스트, 개설, 2, ., 일간, 워스트, 개설, 첫날, 트래픽...   
14       False                                          [ㅈ, ㄱ, ㄴ]   
15       False  [민영화, 했, 습니다, 로, 고쳐, 주, 세여, 뿌, 잉, 뿌, 잉, 는, 공지,...   
16       False  [여러분, 의, 치킨, 은, 양념, 하, 십니까, 아침, 부터, 치킨, 이, 땡겨,...   
17       False  [아이디, 뭘로, 가입, 했, 는지, 기억, 이, 나, 질, 않, 아서, 고민, 입...   
18       False                                               [ㅠㅠ]   
19       False  [아이디, 는, 받, 아, 놓, 고, 로그인, 은, 이메일, 로, 함, ., ., ...   
20       False  [애국, 진보, 인가, ㅂ, ㅅ, 아치, 까, 는, 대신, 에, 눈, 새, 상마, ...   
21       False                                [쏘, 워, 츙워츙워츙, 원, ㅌ]   
22       False  [', 기행, 천지, ', 일, 베, …, 편의점, 호빵, ,, 담배, 와, 같이,...   
23       False  [약속, 나가, 려다가, 번뜩, 떠올라서, 후딱, 베, 낌, 맘, 에, 드, 시, ...   
24       False                                     [외쳐, !, EE, !]   
25       False                                                 []   
26       False                                [못, 하, 면, ㅁ, ㅇㅎ, .]   
27       False  [http, :, //, bbs, 1, ., agora, ., media, ., d...   
28       False                                [일, 베, 충, 박멸, !, !]   
29       False                                [하, ↘, 지마, 루, 요, ↗]   
...        ...                                                ...   
172014   False  [얼마, 전, 우리, 빌라, 의, 귀염둥이, 보더, 콜리, 장군, 이, 가, 없, ...   
172015    True  [머리, 가, 다, 빠져, 버렸, 는데, 머리, 다시, 자라, 게, 하, 는, 방법...   
172016   False                                                 []   
172017   False  [전, 아이돌, 성장, 물, 을, 안, 좋아하, 는, 편, 입니다, ., 뻔하, 잖...   
172018   False  [저, 사진, 넣, 으려니, 지원, 하, 지, 않, 는, 파일, 이, 라는데, .,...   
172019   False  [아까, 부터, 제, 가, 쓴, 글, 이, 밭, 에서, 보이, 지, 도, 않, 고,...   
172020   False     [www, ., ilwar, ., com, /, fashion, /, 189620]   
172021   False  [트롤, 밭, 게시물, 이, 보이, 기, 시작, 합니다, ., .., http, :...   
172022   False  [http, :, //, www, ., segye, ., com, /, conten...   
172023   False  [친구, 가, 다른, 친구, 하, 고, 같이, 있, 으면, ,, 바로, 카톡, 으로...   
172024   False  [그, 중, 우리, 나라, 에서, 독재, 한, 분, 들, 또는, 논란거리, 를, 마...   
172025   False  [무서웠, 습니다, ., ., 마치, 마약, 금단, 증세, +, 외로움, +, 소통...   
172026   False                                            [비, 연시]   
172027   False  [잘, 쓰, 던, 아이패드, 를, 팔, 고, (, 두, 달, 전, 중고, 로, 샀,...   
172028   False  [어떻게, 알, 고, 친, 추, 를, 하, 신지, 모를, 한, 농부, 님, 과, 서...   
172029   False  [이르, 지만, 작별, 하, 고, 싶, 네요, ., 2014, 년, 얼마, 남, 지...   
172030   False  [올해, 도, 이제, ., .., 26, 시간, 여, 남, 았, 구, 농, ., ....   
172031   False  [겜, 덕인, 나농, 이, 마우스, 를, 바꾸, 려고, 하, 는데, 게이밍, 마우스...   
172032   False  [오늘, ', 이강윤, 의, 오늘, ', 에서, 우원식, 의원, 이, 인터뷰, 도,...   
172033   False                    [카피, 문구, 어디, 서, 본, 듯, 한데, ., .]   
172034   False  [때문, 에, 땅콩, 부사장, 님, 은, 곧, 수감, 될, 예정, 입니다, 이상, ...   
172035   False  [...., 라고, 쓰, 고, 농약, 짤, 줍, 용, 글, 이, 라고, 읽, 는다,...   
172036   False  [국내, 최고, 의, 팟캐스트, ", 그것, 은, 알, 기, 싫, 다, ", 의, ...   
172037   False  [허접, 한, 실력, 이, 지만, 개발자, 가, 직업, 인지라, 보, 고, 있, 는...   
172038   False  [무한도전, 의, 기획력, 에, 감탄, 하, 는, 것, 이, 한, 두, 번, 이, ...   
172039   False  [분명, 어데, 부딪치, 지, 도, 않, 았, 는데, 멍, 이, 들었, 농, ., ...   
17204

### predictor, model 생성

In [6]:
from sklearn.ensemble import RandomForestClassifier

label = 'istroll'
pre = labeled_train.columns.drop(['text_morphs', label])

model = RandomForestClassifier(n_estimators=10, n_jobs=6)

###cross validation

In [7]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, labeled_train[pre], labeled_train[label], cv=cv_value, scoring="roc_auc")
cv_result = scores.mean()

print(cv_result)

0.637526911627


##Result
  * text의 morphs를 bag of words로 변환 (feature : 1,000개) - 0.637526911627

In [8]:
import pickle

pickle.dump(labeled_train[pre], open("text_%d.p" % max_features, "wb"), protocol = pickle.HIGHEST_PROTOCOL)